Libraries/LibJS/Runtime/JSONObject.cpp

```diff
@@ -0,0 +1,484 @@
+        for (auto i = 0; i < gap_size; ++i)
+            gap_builder.append(' ');
+        state.gap = gap_builder.to_string();
+    } else if (space.is_string()) {
+        auto string = space.as_string().string();
+        if (string.length() <= 10) {
+            state.gap = string;
+        } else {
+            state.gap = string.substring(0, 10);
+        }
+    } else {
+        state.gap = String::empty();
+    }
+
+    auto* wrapper = Object::create_empty(interpreter, interpreter.global_object());
+    wrapper->define_property(String::empty(), value);
+    if (interpreter.exception())
+        return {};
+    auto result = serialize_json_property(interpreter, state, String::empty(), wrapper);
+    if (interpreter.exception())
+        return {};
+    if (result.is_null())
+        return js_undefined();
+    return js_string(interpreter, result);
+}
+
+String JSONObject::serialize_json_property(Interpreter& interpreter, StringifyState& state, const PropertyName& key, Object* holder)
+{
+    auto value = holder->get(key);
+    if (value.is_object()) {
+        auto to_json = value.as_object().get("toJSON");
+        if (interpreter.exception())
+            return {};
+        if (to_json.is_function()) {
+            MarkedValueList arguments(interpreter.heap());
+            arguments.append(js_string(interpreter, key.to_string()));
+            value = interpreter.call(to_json.as_function(), value, move(arguments));
+            if (interpreter.exception())
+                return {};
+        }
+    }
+
+    if (state.replacer_function) {
+        MarkedValueList arguments(interpreter.heap());
+        arguments.values().append(js_string(interpreter, key.to_string()));
+        arguments.values().append(value);
+        value = interpreter.call(*state.replacer_function, holder, move(arguments));
+        if (interpreter.exception())
+            return {};
+    }
+
+    if (value.is_object()) {
+        auto& value_object = value.as_object();
+        if (value_object.is_number_object() || value_object.is_boolean_object() || value_object.is_string_object() || value_object.is_bigint_object())
+            value = value_object.value_of();
+    }
+
+    if (value.is_null()) {
+        return "null";
+    } else if (value.is_boolean()) {
+        return value.as_bool() ? "true" : "false";
+    } else if (value.is_string()) {
+        return quote_json_string(value.as_string().string());
+    } else if (value.is_number()) {
+        if (value.is_finite_number())
+            return value.to_string(interpreter);
+        return "null";
+    } else if (value.is_object() && !value.is_function()) {
+        if (value.is_array())
+            return serialize_json_array(interpreter, state, static_cast<Array&>(value.as_object()));
+        return serialize_json_object(interpreter, state,  value.as_object());
+    } else if (value.is_bigint()) {
+        interpreter.throw_exception<TypeError>(ErrorType::JsonBigInt);
+    }
+    return {};
```

Comment:
While the "return {};" here works, I think a "return {};" after the BigInt and an ASSERT_NOT_REACHED(); at the end would be clearer (unless it passes through the if/else in other handled cases ?)

Reply:
Hmm. Rereading the code, it seems like there is values not explicitely handled which should be passing through and just be silently swallowed (namely, functions and symbols). So the `ASSERT_NOT_REACHED();` wouldn't make sense, but adding an extra `return {};` after `interpreter.throw_exception()` could still make it less surprising / more future-proof.

Reply:
Yes, this is intentional; Symbols and Functions reach the end of the function, as they are ignored

Reply:
Ehh, not sure what use an additional `return {}` would have when the very next statement after the if statement is a `return {}`.

Reply:
It's mostly for readability sake, at this point, so feel free to let everything as-is. I just had to think about it for a bit while reading this piece of code, and having an additional useless `return {};` would have made the overall behaviour more obvious ^^